torch/_utils.py

```diff
@@ -177,7 +177,21 @@ def _get_async_or_non_blocking(function_name, non_blocking, kwargs):
 # be a TypedStorage
 def _rebuild_tensor(storage, storage_offset, size, stride):
     # first construct a tensor with the correct dtype/device
-    t = torch.empty((), dtype=storage.dtype, device=storage._untyped_storage.device)
+    dtype = storage.dtype
+    device = storage._untyped_storage.device
+    # BEGINNING OF THIAGO'S HACK
```

Comment:
I'm not exactly sure, but to me this seems like a valid way to do it, not just a hack. I'm not sure why it would be better to do this in `torch/serialization.py`, and doing it there would require duplication.

I think all that matters is that at some point the storage has to become `device=meta`, and at some point the tensor has to become `device=device`. It seems fine to do that right when it's being rebuilt, as opposed to afterward, but I could be missing something

Reply:
More than glad to keep this solution and remove the conversation starters comments :)

Reply:
The main problem with this approach is, by this point, you've already spent time useless deserializing the true storage into memory, only to throw it out and then replace it with a meta.

Reply:
Any suggestions on how to go about this?

For the repro

```python
from torch._subclasses import fake_tensor
import transformers

fake_mode = fake_tensor.FakeTensorMode(allow_non_fake_inputs=False)
with fake_mode:
    fake_model = transformers.AutoModel.from_pretrained("sshleifer/tiny-gpt2") 
```
I removed the `_rebuild_tensor` hack from this PR and added the following patches at `torch/serialization.py::_legacy_load`

```bash
 def _validate_privateuse1_device(location, backend_name):
@@ -1132,6 +1132,8 @@ def _legacy_load(f, map_location, pickle_module, **pickle_load_args):
                     args = pickle_module.load(f, **pickle_load_args)
                     key, location, storage_type = args
                     dtype = storage_type._dtype
+                    if torch._guards.detect_fake_mode(None) is not None:
+                        location = 'meta'
                     obj = cast(Storage, torch.UntypedStorage)._new_with_file(f, torch._utils._element_size(dtype))
                     obj = restore_location(obj, location)
                     # TODO: Once we decide to break serialization FC, we can
@@ -1191,8 +1193,8 @@ def _legacy_load(f, map_location, pickle_module, **pickle_load_args):
         elif typename == 'storage':
             storage_type, root_key, location, numel, view_metadata = data
             # THIAGO: I am sure the fix should be here
+            if torch._guards.detect_fake_mode(None) is not None:
+                location = 'meta'
             location = _maybe_decode_ascii(location)
             dtype = storage_type.dtype
```

The error will be

```bash
> /opt/pytorch/torch/serialization.py(1265)_legacy_load()
-> magic_number = pickle_module.load(f, **pickle_load_args)
(Pdb) c
Traceback (most recent call last):
  File "/opt/conda/envs/ptca/lib/python3.11/site-packages/transformers/modeling_utils.py", line 519, in load_state_dict
    return torch.load(checkpoint_file, map_location=map_location)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/serialization.py", line 1039, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/serialization.py", line 1275, in _legacy_load
    result = unpickler.load()
             ^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_utils.py", line 216, in _rebuild_tensor_v2
    tensor = _rebuild_tensor(storage, storage_offset, size, stride)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_utils.py", line 195, in _rebuild_tensor
    return t.set_(storage._untyped_storage, storage_offset, size, stride)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 871, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1207, in dispatch
    return self._cached_dispatch_impl(func, types, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 949, in _cached_dispatch_impl
    output = self._dispatch_impl(func, types, args, kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1424, in _dispatch_impl
    r = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_ops.py", line 571, in __call__
    return self_._op(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempted to set the storage of a tensor on device "meta" to a storage on different device "cpu".  This is no longer allowed; the devices must match.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/envs/ptca/lib/python3.11/site-packages/transformers/modeling_utils.py", line 523, in load_state_dict
    if f.read(7) == "version":
       ^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/pytorch/gh105077.py", line 8, in <module>
    fake_model = transformers.AutoModel.from_pretrained("sshleifer/tiny-gpt2")  # raises OSError: Unable to load weights from pytorch checkpoint file for '...' at  If you tried to load a PyTorch model from a TF 2.0 checkpoint, please set from_tf=True.
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/ptca/lib/python3.11/site-packages/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    return model_class.from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/ptca/lib/python3.11/site-packages/transformers/modeling_utils.py", line 3381, in from_pretrained
    state_dict = load_state_dict(resolved_archive_file)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/ptca/lib/python3.11/site-packages/transformers/modeling_utils.py", line 535, in load_state_dict
    raise OSError(
OSError: Unable to load weights from pytorch checkpoint file for '/root/.cache/huggingface/hub/models--sshleifer--tiny-gpt2/snapshots/5f91d94bd9cd7190a9f3216ff93cd1dd95f2c7be/pytorch_model.bin' at '/root/.cache/huggingface/hub/models--sshleifer--tiny-gpt2/snapshots/5f91d94bd9cd7190a9f3216ff93cd1dd95f2c7be/pytorch_model.bin'. If you tried to load a PyTorch model from a TF 2.0 checkpoint, please set from_tf=True.
```

Reply:
It seems like at least one problem in the case of `_legacy_load` is that the data is still being loaded from the file, with the `cast(Storage, torch.UntypedStorage)._new_with_file` call, and then the `restore_location` call tries to move that real data onto the meta device.

I think the `_new_with_file` call needs to be avoided and instead create an `UntypedStorage` with meta device. For instance, with `torch.UntypedStorage(nbytes, device='meta')`. I'm not sure at the moment how to get the number of bytes from the file `f`, but it should be possible to get that information without loading the whole file into a temporary storage

Reply:
To unblock me, would be possible to merge this solution with a TODO: so that we can keep working on a more efficient implementation? I am stuck on this pretty bad =/

Reply:
> It seems like at least one problem in the case of `_legacy_load` is that the data is still being loaded from the file, with the `cast(Storage, torch.UntypedStorage)._new_with_file` call, and then the `restore_location` call tries to move that real data onto the meta device.
> 
> I think the `_new_with_file` call needs to be avoided and instead create an `UntypedStorage` with meta device. For instance, with `torch.UntypedStorage(nbytes, device='meta')`. I'm not sure at the moment how to get the number of bytes from the file `f`, but it should be possible to get that information without loading the whole file into a temporary storage

what confuses me the most is the jump from `unpickler.load` to `rebuild_tensor`.  Something on this unpickler's load is creating a `TypedStorage` with `cpu` device while the code at `unpickler.persistent_load` was changed to use `meta` device

Reply:
https://github.com/pytorch/pytorch/pull/119584/files is what I am trying as an alternative to changing `_rebuild_tensor`

Reply:
As for merging this with a TODO, it's not really my call. But I feel like it would be best to get it working the right way. You'll have to wait for the next release to get access to the fix anyway, unless you're using the nightly build or building from source.

The jump from `unpickler.load` to `_rebuild_tensor`, `_rebuild_tensor_v2`, or `_rebuild_tensor_v3` happens because at the time when the tensor was saved, during `torch.save`, [`Tensor.__reduce_ex__`](https://github.com/pytorch/pytorch/blob/d9d8c2b79fd8dce7226f9157c5a67adb6d30feff/torch/_tensor.py#L212) gets called, and that method defines how the tensor should get converted into serial data, including the name of the function (`_rebuild_tensor*`) that should be called to rebuild the tensor from the data that was loaded from the file. In most cases, [`Tensor._reduce_ex_internal`](https://github.com/pytorch/pytorch/blob/d9d8c2b79fd8dce7226f9157c5a67adb6d30feff/torch/_tensor.py#L247) is called during the `torch.save` call, and the following is the most important part of that function:
https://github.com/pytorch/pytorch/blob/d9d8c2b79fd8dce7226f9157c5a67adb6d30feff/torch/_tensor.py#L408-L436

Notice that the `rebuild_func` is normally set to `_rebuild_tensor_v2`, in the typical case. `args` contains the tensor's storage, along with other info about the tensor.

During the `torch.save` call, the pickler will serialize all of those args and write them to the file--all of that information written to the file represents the tensor. The pickler also reduces each of those args in a similar way until it is left with just byte arrays that it can write to the file.

So when we `torch.load` a file that contains the serialized data for a tensor, during the `unpickler.load` call, each of the `args` that the tensor was reduced to gets deserialized, and then whatever `rebuild_func` was written to the file during `torch.save` is called with those args.

The serialization and deserialization of storages is implemented a little differently than other objects. Instead of using a `__reduce_ex__` function, the `pickler.persistent_id` functions [here](https://github.com/pytorch/pytorch/blob/d9d8c2b79fd8dce7226f9157c5a67adb6d30feff/torch/serialization.py#L839) and [here](https://github.com/pytorch/pytorch/blob/d9d8c2b79fd8dce7226f9157c5a67adb6d30feff/torch/serialization.py#L768) are used, so that if two tensors share the same storage, the storage is only written to the file once. And then during `torch.load`, the corresponding `persistent_load` function is used as well.

Hopefully that gives you a better picture of what's going on here